### PR TITLE
fix(agent): keep StorageACK retries collector-owned (#1577)

### LINKED ISSUE #1577: bug: StorageACK requests remain in durable outbox after ACKCollector terminates
## Bug Description

StorageACK collection is a synchronous request/response quorum operation, but failed StorageACK requests are also persisted in the Universal Messenger sender outbox for up to 24 hours. Once the live `ACKCollector` reaches quorum, times out, or determines quorum is impossible, a later outbox delivery cannot return its ACK to that finished collector or affect the completed publish.

This creates a contract mismatch:

1. `ACKCollector` already owns bounded transport and transient-decline retries.
2. `createACKSendP2P()` uses `messenger.sendReliable()` underneath.
3. A recoverable send failure is persisted in `protocol_outbox`; `createACKSendP2P()` reports it to the collector as a transport failure.
4. The collector retries or finishes independently, but the durable row remains.
5. Periodic and reconnect-triggered outbox drains replay the old PublishIntent after the original quorum operation has ended.

The replay may still cause a core to stage data and compute an ACK, but there is no active collector to consume the response. If eventual replication to additional cores is desired, it needs an explicit finalization-aware one-way contract rather than an orphaned request/response retry.

## Expected Behavior

- StorageACK transport retries must have one clear owner.
- A synchronous collector should retain reliable-envelope encoding and receiver-side idempotency without leaving sender-outbox rows after the collector ends.
- Alternatively, a persisted ACK request must remain attached to resumable publish/quorum state that can consume the eventual response; that state does not exist for the current synchronous publish path.
- Reconnect-triggered drains must not bypass backoff repeatedly for every libp2p `connection:open` event.
- Outbox draining must be globally bounded and non-reentrant.

## Actual Behavior

On one edge-node soak:

- `protocol_outbox` contained 2,648 rows, all `/dkg/10.0.1/storage-ack`.
- The rows targeted 23 peers and accumulated more than 1.06 million attempts in about 23 hours.
- Individual messages reached more than 1,100 attempts despite the configured backoff ladder ending at two hours.
- `processOutboxOnConnect()` intentionally retries every pending row for a peer even when `nextAttemptAt` is still in the future.
- The 30-second periodic tick has no drain-wide in-flight guard or batch/concurrency limit, so a slow drain can overlap later ticks.
- Most terminal errors were aborted sends, identity-probe timeouts, protocol negotiation failures, or relay reservation failures.

The table itself was only about 1.1 MB. The impact is repeated dials, network admission probes, core-side request handling, SQLite SELECT/UPDATE/prepare work, garbage collection, and log amplification.

## Relevant Logs / Error Output

```text
protocol=/dkg/10.0.1/storage-ack rows=2648 attempts=1064376 peers=23
highest per-message attempts: 1110
examples:
Network identity probe failed ... The operation was aborted due to timeout
Protocol selection failed - could not negotiate /dkg/10.0.1/network-identity
This operation was aborted
```

## Steps to Reproduce

1. Start an edge node with multiple connected/discovered core peers and durable Messenger stores.
2. Publish repeatedly while some candidate peers are unreachable, foreign-network, slow, or missing the StorageACK protocol.
3. Allow the collector to reach quorum or terminate.
4. Query `protocol_outbox` for `/dkg/10.0.1/storage-ack` rows.
5. Leave the node running through relay/connection churn.
6. Observe attempts increasing far faster than the configured backoff ladder permits.

## Possible Root Causes

- `packages/agent/src/dkg-agent.ts:createACKSendP2P()` calls `sendReliable()` for a synchronous request/response protocol.
- `packages/agent/src/p2p/messenger.ts:sendReliable()` unconditionally persists recoverable failures.
- `packages/agent/src/p2p/messenger.ts:processOutboxOnConnect()` ignores `nextAttemptAt` and drains all rows for every connection-open event.
- `packages/agent/src/dkg-agent-lifecycle.ts` starts `processOutboxTick()` from `setInterval` without preventing overlapping drain executions.
- `SqliteProtocolOutboxStore` prepares SELECT/UPDATE statements on each retry, amplifying the cost once attempts become excessive.

## DKG daemon / version

`10.0.5`, `testnet-canary` commit `8d282fc2326cb66161c33d6a7ae46154222ad396`

## Surface / Component

StorageACK collector, Universal Messenger outbox, SQLite protocol outbox

## Operating System

macOS 26.5.1

## Node.js Version

22.22.0

## Environment — endpoints, config, health checks

- Edge node on Base Sepolia (`base:84532`)
- 40-publish measured workload plus prior publish tests
- Durable SQLite Messenger outbox enabled
- Significant relay and network-admission churn

## Suggested Investigation

1. Add a `persistOnFailure: false` / `queueOnFailure: false` reliable request mode, or a dedicated `sendReliableRequest()` API, preserving envelope and receiver dedup while leaving retry ownership with `ACKCollector`.
2. Apply it to StorageACK and audit VerifyCollector for the same request/response mismatch.
3. If post-quorum replication is required, define a separate idempotent, finalization-aware message whose response is not needed by a dead caller.
4. Independently harden the generic outbox: first-connection transition instead of every connection-open event, honor due time/cooldown, one drain lock, bounded batch/concurrency, and per-peer/protocol row/byte limits.
5. Cache prepared SQLite statements.

## Acceptance Criteria

- A failed or quorum-abandoned StorageACK request leaves no orphaned sender-outbox row after the collector terminates.
- Existing collector transport retries and transient-decline retries remain intact.
- Reliable envelope decoding and receiver-side idempotency remain intact.
- A 100-publish mixed-health soak does not accumulate stale StorageACK rows or attempts after publishes settle.
- Reconnect churn cannot bypass the configured retry policy or start overlapping unbounded drains.
- Tests cover quorum success with slow peers, quorum failure, daemon restart, and repeated connection-open events.

## Notes / Related Issues

- Related network isolation history: #1543.
- Related prior performance incident: #939.
- Found during a `testnet-canary` QA/soak pass.
- Existing issues were searched; no duplicate was found.

### PR DESCRIPTION
**Pull request by branarakic**
## Summary

- add an explicit `queueOnFailure: false` mode to reliable request sends
- use it for publish and update StorageACK transports
- preserve ReliableEnvelope framing and receiver idempotency
- return recoverable failures to ACKCollector without creating durable sender rows

Closes #1577

## Verification

- `pnpm --filter @origintrail-official/dkg-agent run build`
- added Messenger outbox-state and ACK transport wiring regressions (local agent Vitest process did not terminate cleanly; CI will run them)

### CODE REVIEW

File: packages/agent/src/p2p/messenger.ts [hunk 1]
```diff
@@ -669,6 +674,9 @@ export class Messenger {
       if (!isRecoverableMessengerSendError(err, errMsg)) {
         throw err;
       }
+      if (opts.queueOnFailure === false) {
+        throw err;
```
Comment (otReviewAgent) [inline packages/agent/src/p2p/messenger.ts:703]: 🟡 Issue: Non-queued recoverable failures leave SLO attempt state behind

**What's wrong**
The new branch disables durable queueing but still exits through a path that was designed for queued or exceptional failures. `sendReliable` records SLO attempt state before dialing; when a recoverable transport error is queued, that state is later cleared by delivery or outbox expiry. With `queueOnFailure:false`, the error is thrown and no outbox entry exists, so common recoverable failures can accumulate stale state for every generated messageId.

**Example**
A StorageACK send to a peer with `no valid addresses for peer` now calls `sendReliable(..., { queueOnFailure: false })`. Each ACKCollector retry gets a fresh generated messageId, stores a `firstAttemptAt` entry, then throws from line 678. Because no outbox entry exists, `dropExpiredOutbox()` never removes those keys, so repeated unreachable ACK peers grow this map indefinitely.

**Suggested direction**
Clear the `firstAttemptAt` entry on the non-queued recoverable-failure path before rethrowing, since there is no durable outbox entry that can later deliver or expire it.

**For Agents**
In `Messenger.sendReliable`, handle the `queueOnFailure === false` recoverable-error branch by clearing the SLO key for `(protocolId, peerId, messageId)` before throwing, while preserving the existing queued path’s behavior. Add a focused test that a recoverable failure with `queueOnFailure:false` does not leave pending SLO state, ideally via exposed stats or a test-visible hook.

File: packages/agent/src/p2p/messenger.ts [hunk 2]
```diff
@@ -184,6 +184,11 @@ export interface SendReliableOpts {
    * (e.g. join-approval).
    */
   maxAgeMs?: number;
+  /**
+   * Disable durable sender retries for bounded request/response flows whose
+   * caller owns retry and response lifetime (for example StorageACK).
+   */
+  queueOnFailure?: boolean;
```
Comment (otReviewAgent) [inline packages/agent/src/p2p/messenger.ts]: 🟡 Issue: Do not hide a second reliability mode behind an optional boolean

**What's wrong**
`sendReliable` used to have one coherent contract: recoverable send failures become durable outbox entries. This patch makes the same public method silently switch to a different contract when `queueOnFailure` is false, but the method name, result type, and surrounding docs still describe the durable API. That adds mode branching to a large shared module and forces feature-specific knowledge from StorageACK into the generic messenger surface. Over time, this kind of boolean mode tends to spread because callers must remember which combination returns `queued`, throws, updates outbox/SLO state, or participates in retries.

**Example**
StorageACK now calls sendReliable with `{ timeoutMs, queueOnFailure: false }`, which means recoverable transport failures throw instead of returning the documented `{ delivered: false, queued: true }` shape. Other callers use the same method name and return type but get durable outbox semantics by omission.

**Suggested direction**
Prefer splitting the shared mechanics from the durability policy. A small internal helper for envelope/idempotency/router-send could be reused by `sendReliable` and an explicit request-owned method, avoiding a broad `queueOnFailure?: boolean` knob on every reliable caller. If keeping one method, use a named discriminated policy such as `failurePolicy: 'queue' | 'throw'` with typed result differences rather than an optional negative boolean.

**For Agents**
Look at `Messenger.sendReliable` and the ACK adapter in `packages/agent/src/dkg-agent.ts`. Preserve reliable envelope/idempotency behavior for ACK sends, but separate the retry policy boundary: either extract a lower-level framed request/response primitive used by ACK and keep `sendReliable` as the durable-outbox API, or make the policy an explicit discriminated mode/overload so the return/throw contract is visible. Existing durable callers should still enqueue recoverable failures; ACK should still keep failures collector-owned.

Comment (otReviewAgent) [reply packages/agent/src/p2p/messenger.ts]: 🟡 Issue: Avoid turning durable sendReliable into a policy-switched API

**What's wrong**
`sendReliable` is the canonical durable substrate API, but this change makes durability optional through a caller-supplied flag. That adds a second ownership model to an already large method and forces readers to hold the option union, overloads, SLO cleanup exception, and ACK call-site intent together. A named request-owned path would make the boundary explicit and delete the ad-hoc branch from the durable flow.

**Example**
The ACK path asks for `sendReliable(..., { failurePolicy: 'throw' })`, so the same public method now means either durable queueing or request-owned retry depending on an option literal. The only thing separating those contracts is the branch at line 702.

**Suggested direction**
Split the concept instead of adding a mode flag: keep `sendReliable` as the durable queueing API, extract the shared envelope/dedupe/in-flight/SLO attempt into a private helper, and give the ACK path a named throwing/request-owned entry point.

**Confidence note**
This is a structural recommendation rather than a behavioral defect; it is based on the new mode flag being used by a single ACK path in the provided diff.

**For Agents**
Look in `packages/agent/src/p2p/messenger.ts` and `packages/agent/src/dkg-agent.ts`. Preserve the ACK behavior: reliable envelope/idempotency framing, recoverable transport failures thrown to the request-owned retry loop, no outbox persistence, and no leaked SLO marker. Prefer extracting the common framed wire-attempt logic and exposing a named request-owned method/helper instead of a policy flag on durable `sendReliable`; update ACK wiring and existing sendReliable tests to prove both public contracts remain distinct.

File: packages/agent/test/v10-ack-provider-wiring.test.ts [hunk 3]
```diff
@@ -285,9 +285,11 @@ describe('DKGAgent.createV10ACKProvider — structured ACK verifier wiring (PR #
 
     expect(sendReliable).toHaveBeenNthCalledWith(1, 'peer-a', '/dkg/test/storage-ack', payload, {
       timeoutMs: 60_000,
+      queueOnFailure: false,
```
Comment (otReviewAgent) [inline packages/agent/test/v10-ack-provider-wiring.test.ts]: 🔴 Bug: Keep ACK wiring tests on the same option contract as production

**What's wrong**
The PR introduces `failurePolicy: 'throw'` as the production API but the changed tests encode `queueOnFailure: false`. This creates two names for the same concept and weakens the test as living documentation of the boundary. Even if the implementation direction changes, the production call and test should share one typed contract.

**Example**
The production ACK sender calls `sendReliable(peer, protocol, payload, { timeoutMs, failurePolicy: 'throw' })`, while the wiring spec asserts `queueOnFailure: false`. That is a different option contract, and the untyped `vi.fn` mock lets the test drift away from the production interface.

**Suggested direction**
Replace the stale `queueOnFailure: false` assertions with the actual `failurePolicy: 'throw'` contract, and give the mock a type tied to the production ACK messenger boundary instead of leaving it as a shape-less `vi.fn`.

**For Agents**
Update `packages/agent/test/v10-ack-provider-wiring.test.ts` around the ACK sendReliable mocks/assertions. Preserve the behavior that ACK sends do not queue recoverable failures. Type the mock against the ACK messenger/sendReliable contract or import the shared option/result types so the test cannot assert a stale option name again.

ACK wiring tests still assert the old queue option

**What's wrong**
The changed behavior is that ACK sends opt into the explicit throw policy, but the updated wiring tests assert an option name that production no longer sends. That makes the validation evidence stale and likely breaks the test suite instead of proving ACK sends no longer enqueue recoverable failures.

**Example**
The first ACK wiring assertion now expects `sendReliable(..., { timeoutMs: 60000, queueOnFailure: false })`, but the changed implementation calls `sendReliable(..., { timeoutMs: 60000, failurePolicy: 'throw' })`. A regression that drops `failurePolicy: 'throw'` from ACK sends would not be verified by these assertions as written.

**Suggested direction**
Replace the added `queueOnFailure: false` expectations with assertions for the new `failurePolicy: 'throw'` contract so the tests verify the behavior this PR introduces.

**For Agents**
Update `packages/agent/test/v10-ack-provider-wiring.test.ts` ACK wiring expectations to assert `failurePolicy: 'throw'` wherever ACK sends are expected to be request-owned. Also update the failed-send expectation to match the new in-flight/error behavior where applicable, then run the affected Vitest file.

File: packages/agent/src/p2p/messenger.ts (hunk 1 above)
Comment (otReviewAgent) [reply packages/agent/src/p2p/messenger.ts:703]: 🔴 Bug: Request-owned sends skip DHT recovery for address failures

**What's wrong**
The new request-owned path avoids durable ACK outbox rows, but it also bypasses the recovery side effect that `sendReliable` used for stale or missing peer addresses. ACK collection owns the retry loop now, yet those retries no longer refresh peer routing for recoverable address errors, so a temporarily recoverable ACK peer can become a terminal transport failure.

**Example**
If an ACK peer returns `all multiaddr dials failed` or `no valid addresses`, `sendRequestOwned` now throws and the collector retries the peer, but no DHT walk is scheduled to refresh stale/missing peer addresses. The retries can repeat against the same stale routing state until that peer is marked `TRANSPORT_ERROR`, reducing the chance of reaching ACK quorum.

**Suggested direction**
Trigger the same peer-resolution side effect for request-owned recoverable failures without enqueueing a durable outbox row.

**Confidence note**
The local code shows the new request-owned branch exits before the existing DHT recovery hook. I did not run a live devnet to measure quorum impact.

**For Agents**
In `packages/agent/src/p2p/messenger.ts`, keep `sendRequestOwned` non-durable, but preserve address recovery before throwing in the `!queueRecoverableFailure` branch. Add a focused test proving a DHT-triggering recoverable error calls `resolvePeer` while leaving the outbox empty.

File: packages/agent/src/p2p/messenger.ts (hunk 2 above)
Comment (otReviewAgent) [reply packages/agent/src/p2p/messenger.ts]: 🟡 Issue: Avoid encoding the new send mode as a boolean plus cast

**What's wrong**
This adds a second ownership policy to an already large reliable-send method by threading a raw boolean through the core flow. The important invariant, “request-owned sends never return durable queued results,” is enforced by a cast rather than the type boundary. That makes the method harder to evolve safely and hides the exact place where outbox ownership differs.

**Example**
A future edit can add or move a queued return inside `sendFramed`; the `sendRequestOwned` public contract would still compile because the cast at the call site asserts that queued results are impossible instead of proving it through the implementation type.

**Suggested direction**
Split the shared mechanics from the recoverable-failure policy. For example, have a common framed send-attempt helper and let `sendReliable` supply the enqueue policy while `sendRequestOwned` supplies the throw policy, with overloads or a discriminated mode making queued results statically impossible for the request-owned path.

**For Agents**
Refactor `packages/agent/src/p2p/messenger.ts` around `sendReliable`, `sendRequestOwned`, and `sendFramed`. Preserve envelope framing, idempotency, in-flight handling, and request-owned recoverable failures throwing without durable outbox rows. Replace the boolean mode plus cast with an explicit failure policy/overload/discriminated internal helper so the `sendRequestOwned` return type is derived rather than asserted; keep tests proving durable sends queue and request-owned sends do not.

File: scripts/devnet-test-issue-1577-ack-outbox-ownership.sh [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  [[ "$(code_of "$(api "$n" GET /api/status)")" == 200 ]] || fail "node$n is not ready (need 4 cores + edge)"
+done
+[[ -f "$DB" ]] || fail "publisher database missing: $DB"
+
+count_ack_rows() {
+  DB="$DB" node --input-type=module <<'NODE'
+import Database from 'better-sqlite3';
+const db = new Database(process.env.DB, { readonly: true });
+const row = db.prepare(`SELECT COUNT(*) AS n FROM protocol_outbox
+  WHERE protocol IN ('/dkg/10.0.1/storage-ack','/dkg/10.0.2/storage-ack','/dkg/10.0.1/storage-update-ack')`).get();
```
Comment (otReviewAgent) [inline scripts/devnet-test-issue-1577-ack-outbox-ownership.sh:33]: 🟡 Issue: Use canonical protocol constants in the devnet regression script

**What's wrong**
The script duplicates versioned storage ACK protocol IDs instead of reading them from the canonical constants. This creates a small but real maintenance fork in exactly the area the script is meant to guard.

**Example**
If storage ACK moves to another wire version, the agent can update the core constants while this regression script keeps querying the old protocol strings and reports a clean outbox for the wrong protocols.

**Suggested direction**
Avoid hardcoding versioned protocol IDs in the SQL filter. Reuse a shared shell helper for constants or pass the canonical values from the inline Node code into the prepared statement.

**Confidence note**
This is a devnet script, so the operational blast radius is lower than production code, but the duplicated constants are versioned protocol IDs that already have canonical definitions.

**For Agents**
Update `scripts/devnet-test-issue-1577-ack-outbox-ownership.sh` to resolve the storage ACK protocol IDs from the canonical core constants, either by moving the existing `protocol_const` helper into `devnet-lib.sh` or by importing the built `@origintrail-official/dkg-core` constants in the inline Node snippet. Preserve the same SQL count semantics.

File: packages/agent/test/v10-ack-provider-wiring.test.ts (hunk 3 above)
Comment (branarakic) [reply packages/agent/test/v10-ack-provider-wiring.test.ts]: Rechecked on current main: the merged #1624 head already uses sendRequestOwned consistently in production and the wiring test, so the reported option-name mismatch is no longer present. #1658 adds a stronger typed propagation regression proving an address failure travels back through the ACK collector sendP2P boundary with the exact timeout while remaining collector-owned.

File: packages/agent/src/p2p/messenger.ts (hunk 1 above)
Comment (branarakic) [reply packages/agent/src/p2p/messenger.ts:703]: Addressed in #1658 (9d4c271c5). Request-owned recoverable address failures now clear their SLO marker, trigger the existing rate-limited DHT peer-resolution path immediately, and then throw without creating an outbox row. The messenger regression covers resolvePeer, zero outbox/SLO state, and a second collector retry with a fresh message ID remaining outbox-free while DHT recovery is rate-limited. The compiled devnet preflight exercises the same no-valid-addresses path.

File: packages/agent/src/p2p/messenger.ts (hunk 2 above)
Comment (branarakic) [reply packages/agent/src/p2p/messenger.ts]: Addressed in #1658 (9d4c271c5). The raw boolean and return-type cast are gone. sendFramed now takes a named FramedFailurePolicy and has overloads whose request-owned branch returns ThrowingReliableSendResult, so sendRequestOwned derives its non-queued result contract from the implementation rather than asserting it. Focused messenger tests cover both durable queueing and request-owned throwing behavior.

File: scripts/devnet-test-issue-1577-ack-outbox-ownership.sh (hunk 4 above)
Comment (branarakic) [reply scripts/devnet-test-issue-1577-ack-outbox-ownership.sh:33]: Addressed in #1658 (9d4c271c5). protocol_const now lives in scripts/devnet-lib.sh and resolves the canonical core constants, with an optional CORE_CONSTANTS_TS override. The issue-1577 script resolves all three StorageACK IDs through that helper and passes them as parameters to the SQL IN query; it fails early if a constant cannot be resolved.
